scanpipe/pipes/alpine.py

```diff
@@ -20,7 +20,91 @@
 # ScanCode.io is a free software code scanning tool from nexB Inc. and others.
 # Visit https://github.com/nexB/scancode.io for support and download.
 
+import json
+
 from packagedcode import alpine
+from fetchcode.vcs.git import fetch_via_git
+from fetchcode import fetch
+from shutil import copytree
+from scanpipe.models import DiscoveredPackage
+
+APORTS_URL = "https://gitlab.alpinelinux.org/alpine/aports.git"
+APORTS_DIR_NAME = "aports"
+APORTS_SUBDIRS = ["main", "non-free", "community", "testing", "unmaintained"]
+
+
+def download_or_checkout_aports(aports_dir_path, alpine_version, commit_id=None):
+    """
+    Download aports repository and it's branch based on `alpine_version`.
+    Checkout to a branch (alpine version).
+    If `commit_id` is provided also checkout to a commit.
+    Return False if checkout failed otherwise True. #TODO Proper fetchcode patch required (extending #54)
+    """
+    ver = alpine_version.split(".")
+    aports_dir_path = str(aports_dir_path / APORTS_DIR_NAME)
+    fetch_via_git(f"git+{APORTS_URL}@{ver[0]}.{ver[1]}-stable", aports_dir_path)
+    if commit_id:
+        fetch_via_git(f"git+{APORTS_URL}@{commit_id}", aports_dir_path)
+    return True
+
+
+def get_packages_from_db(project):
+    """
+    Get alpine packages from the database that belong to a `project`
+    Yield only root and not yet scanned packages along with paths needed when performing a scan.
+    """
+    for package in DiscoveredPackage.objects.project(project):
+        scan_id = f"{package.name}_{package.version}"
+        scan_target_path = project.tmp_path / scan_id
+        scan_result_path = project.output_path / (scan_id + ".json")
+        if (
+            package.type == "alpine"
+            or (package.source_packages and package.source_packages[0] in package.purl)
+            or not scan_result_path.exists()
+        ):
+            yield scan_target_path, scan_result_path, package
+
+
+def prepare_scan_dir(package_name, scan_target_path, aports_dir_path=None):
+    """
+    Find package's aports path and if found execute the following steps:
+        Copy all the files from that path into `scan_target_path`
+        Download all package's sources into `scan_target_path`
+    The default value of `aports_dir_path` is set to the parent of the `scan_target_path`
+    """
+    if aports_dir_path is None:
+        aports_dir_path = scan_target_path.parent
+    for subdir_name in APORTS_SUBDIRS:
+        apkbuild_dir = aports_dir_path / APORTS_DIR_NAME / subdir_name / package_name
+        if not apkbuild_dir.exists():
+            continue
+        copytree(apkbuild_dir, scan_target_path)
```

Comment:
Why do you need a copy?

Reply:
Because we have to download and extract the sources somwhere  them and it shouldnt be inside of the aports repo directory. Furthermore doing a scan on a single directory is much more readable and elegant.

Reply:
OK, so if I understand correctly you are:
1. making a copy of the aports directory of a given package (which would typically include the APKBUILD and some patches)
2. in this copied directory, you will also fetch the sources (or at least only the remote sources as identified by a URL)
3. finally you will (extract then run a scan of sorts on this directory? )

I think it could be better if you separate each operation and the process could benefit from more documentation.

Reply:
> OK, so if I understand correctly you are:

Yes, exactly.

> I think it could be better if you separate each operation and the process could benefit from more documentation.

If i separated 1. from 2. it would have only `copytree(apkbuild_dir, scan_target_path)` as the function body.

Reply:
So do you want me to split it? Given my reasons above I don't think we should.